### PR TITLE
[V3] Feature/shared type instance cache

### PR DESCRIPTION
**Pull request by FabianTerhorst**
* Its now possible to implement an own caching solution for the type instances

### CODE REVIEW

File: library-core/src/main/java/com/mikepenz/fastadapter/FastAdapter.java [hunk 1]
```diff
@@ -484,14 +507,14 @@ public void registerTypeInstance(Item item) {
...
      */
     public Item getTypeInstance(int type) {
-        return mTypeInstances.get(type);
+        return getTypeInstanceCache().get(type);
     }
 
     /**
      * clears the internal mapper - be sure, to remap everything before going on
      */
     public void clearTypeInstance() {
```
Comment (FabianTerhorst) [inline library-core/src/main/java/com/mikepenz/fastadapter/FastAdapter.java]: This is very critical and maybe should be removed in favor of ```fastAdapter.getTypeInstanceCache().clear();```

File: library-core/src/main/java/com/mikepenz/fastadapter/FastAdapter.java [hunk 2]
```diff
@@ -101,6 +102,29 @@ public FastAdapter() {
...
+        this.mTypeInstanceCache = mTypeInstanceCache;
+    }
+
+    /**
+     *
+     * @return the current type instance cache
+     */
+    public ITypeInstanceCache<Item> getTypeInstanceCache() {
+        if (mTypeInstanceCache == null) {
+            mTypeInstanceCache = (ITypeInstanceCache<Item>) ITypeInstanceCache.DEFAULT;
```
Comment (FabianTerhorst) [inline library-core/src/main/java/com/mikepenz/fastadapter/FastAdapter.java]: Java does not allow generics in static values so we have an warning here.

**Comment by FabianTerhorst** [comment]
Im not sure if the default implementation should be static in this case.

File: library-core/src/main/java/com/mikepenz/fastadapter/FastAdapter.java (hunk 1 above)
Comment (mikepenz) [reply library-core/src/main/java/com/mikepenz/fastadapter/FastAdapter.java]: `getTypeInstanceCache` is anyways just returning the TypeInstanceCache per `FastAdapter` so it'S fine. or do I miss something?

**Pull request merged**